HTTP Methods Examples Provided 20231026

leanneeliatra opened the pull request:
### Description
A customer submitted a query in relation to this page of the documentation:  
https://opensearch.org/docs/latest/api-reference/document-apis/index-document

**Customer query:** _please give a full detailed document with method type. what use put method? like this if describe that can easy to use a developer this apies.-x_

**Answer to customers query:**
In the documentation there is a list of HTTP method examples. The examples are not explained and they cannot be directly inputted to test them.

**Fix**: Add clarification to the documentation. Describe in more detail what the HTTP methods are used for. Give actual examples the user can follow along with in a more stepwise manner, give examples the user can copy paste into the dev tools and run them.

Issues Resolved
This ticket addresses one part of https://github.com/opensearch-project/documentation-website/issues/4314

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

leanneeliatra commented:
@hdhalter @scrawfor99 @Naarcha-AWS, this PR is ready for a review please. Thank you.

scrawfor99 commented on _security/access-control/document-level-security.md:
```diff
@@ -36,6 +36,568 @@ This query specifies that for the role to have access to a document, its `genres
+
+# Document-level security (DLS)
+
+Document-level security lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index permissions** section.
```
Should "permissions" in index permissions be capitalized as a proper noun here?

scrawfor99 commented on _security/access-control/document-level-security.md:
```diff
@@ -36,6 +36,568 @@ This query specifies that for the role to have access to a document, its `genres
+
+Document-level security lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index permissions** section.
+
+![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)
```
@kolchfa-aws can you confirm if the hyphen syntax is correct here?

leanneeliatra replied:
It's a good observation, I was following the page as it was previously but it would be good to get another opinion on what is preferred please @kolchfa-aws

scrawfor99 commented on _security/access-control/document-level-security.md:
```diff
@@ -36,6 +36,568 @@ This query specifies that for the role to have access to a document, its `genres
+  "index_permissions" : [
+    {
+      "index_patterns" : [ "*" ],
+      "allowed_actions" : [ "indices:data/read/search/template" ]---
```
This looks like the formatting is weird?

Naarcha-AWS replied:
A copy and paste error maybe @leanneeliatra?

leanneeliatra replied:
Indeed, thanks

scrawfor99 commented on _security/access-control/document-level-security.md:
````diff
@@ -36,6 +36,568 @@ This query specifies that for the role to have access to a document, its `genres
+}
+```
+
+This query specifies that for the role to have access to a document, its `genres` field must include `Comedy`.
````
Is this supposed to be repeated?

leanneeliatra replied:
That was an error Stephen thanks for catching.

leanneeliatra commented:
Hi all, I have addressed all the comments now for this PR.
To summarise, these are the recent comments and the updated status:
1. _We lost the description of the index, add back in the first 2 sections._ cc @hdhalter 
**This is amended**., see the [updated page](https://github.com/opensearch-project/documentation-website/blob/368f99b916a72df935b12ad8f03142b5f5994dd2/_api-reference/document-apis/index-document.md)
- _We should make sure we follow the API template as per [this page]( https://github.com/opensearch-project/documentation-website/blob/main/templates/API_TEMPLATE.md)_ cc @hdhalter 
I'm not sure how I am currently deviating. Are the current changes okay, or can you inform me of what it is I am missing please, as I’m not certain from reviewing the page, thank you. 
- _We had repeated a section of code here_, @scrawfor99 & @Naarcha-AWS picked up on this, thanks 
**this is amended**.
- I have capitalised the word `Permissions` in `Index Permissions` **amended** cc @scrawfor99 
- The hypen in document-level and field-level can stay, in keeping with the rest of the page. cc @scrawfor99

hdhalter commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+
+## Adding a sample index and documents to use the HTTP methods 
+
+Before you can search for data, you must first add documents.
```
```suggestion
To add a documents, do the following:
```

hdhalter commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+
+Before you can search for data, you must first add documents.
+To do this follow the steps:
+1. Open OpenSearch Dashboards
```
```suggestion
1. Open OpenSearch Dashboards.
```

hdhalter commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+Before you can search for data, you must first add documents.
+To do this follow the steps:
+1. Open OpenSearch Dashboards
+2. Navigate to the hamburger menu 
```
```suggestion
2. Navigate to the hamburger menu. 
```

hdhalter commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+To do this follow the steps:
+1. Open OpenSearch Dashboards
+2. Navigate to the hamburger menu 
+3. In the **Management** section, choose **Dev Tools**
```
```suggestion
3. In the **Management** section, choose **Dev Tools**.
```

hdhalter commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+1. Open OpenSearch Dashboards
+2. Navigate to the hamburger menu 
+3. In the **Management** section, choose **Dev Tools**
+4. Run the following commands in the Dev tools by pressing the green triangle play button to send the request.
```
```suggestion
4. Enter a command, then press the green triangle play button to send the request. Some example commands follow.
```

hdhalter reviewed (commented):
I made a few tweaks. Thanks!

leanneeliatra commented:
> I made a few tweaks. Thanks!

Thanks a million @hdhalter.

We should be good for final review then I think,  @Naarcha-AWS, could you confirm and give a final glance over the changes please?

hdhalter reviewed (commented):
Can we please change the first line? 
From: "Before you can search for data, you must first add documents. This operation adds a single document to your index."
To: "You can use the Index document operation to add a single document to your index."

leanneeliatra commented:
> Can we please change the first line? From: "Before you can search for data, you must first add documents. This operation adds a single document to your index." To: "You can use the Index document operation to add a single document to your index."

Yes thank you @hdhalter, that is now complete.

Naarcha-AWS commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+
+## Adding a sample index and documents to use the HTTP methods 
+
+To add a documents, do the following:
```
```suggestion
If you want to quickly test the Document APIs, you can quickly add a document by doing the following:
```

hdhalter replied:
You probably don't want to use quickly twice in a sentence.

leanneeliatra replied:
Thank you @hdhalter this is now updated.

Naarcha-AWS reviewed (approved):
One small comment then we'll pass it to editorial.

leanneeliatra commented:
> One small comment then we'll pass it to editorial.

That's great @Naarcha-AWS thanks, I have made that small change and addressed all other comments now.
To check also, should we leave in the section `## Adding a sample index and documents to use the HTTP methods` or do you think it's a bit too much information?
Thanks

leanneeliatra commented:
Final changes from sync call & confirmations from @Naarcha-AWS - yet to be made. I will update once these are complete.
1.  Change Hamburger to actions menu. 
2. curl based commands through terminal
3. Comment `should we leave in the section ## Adding a sample index and documents to use the HTTP methods` this section is okay to leave in from Nates review.

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -11,7 +11,7 @@ redirect_from:
 {: .label .label-purple}
 
-Before you can search for data, you must first add documents. This operation adds a single document to your index.
+You can use the Index document operation to add a single document to your index.
```
```suggestion
You can use the index document operation to add a single document to your index.
```

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -11,7 +11,7 @@ redirect_from:
 {: .label .label-purple}
 
-Before you can search for data, you must first add documents. This operation adds a single document to your index.
+You can use the Index document operation to add a single document to your index.
```
Should `index document` be in code font?

leanneeliatra replied:
Yes, thank you Nate

natebower commented on _api-reference/document-apis/index-document.md:
````diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
 POST <index>/_create/<_id>
 ```
 
+- PUT adds or updates documents with a specified ID in the index. Used for controlled document creation or updates.
````
```suggestion
- PUT adds or updates documents in the index with a specified ID. Used for controlled document creation or updates.
```

natebower commented on _api-reference/document-apis/index-document.md:
````diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
 ```
 
+- PUT adds or updates documents with a specified ID in the index. Used for controlled document creation or updates.
+- POST adds documents to the index with auto-generated IDs. Useful for adding new documents without specifying IDs.
````
```suggestion
- POST adds documents with auto-generated IDs to the index. Useful for adding new documents without specifying IDs.
```

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
 
+- PUT adds or updates documents with a specified ID in the index. Used for controlled document creation or updates.
+- POST adds documents to the index with auto-generated IDs. Useful for adding new documents without specifying IDs.
+- `_create` is a type identifier indicating document creation should only occur if the document with the specified ID doesn't already exist.
```
```suggestion
- `_create` is a type identifier indicating that document creation should only occur if the document with the specified ID doesn't already exist.
```

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
+- PUT adds or updates documents with a specified ID in the index. Used for controlled document creation or updates.
+- POST adds documents to the index with auto-generated IDs. Useful for adding new documents without specifying IDs.
+- `_create` is a type identifier indicating document creation should only occur if the document with the specified ID doesn't already exist.
+- `<index>` represents the name of the index where the document will be added.
```
```suggestion
- `<index>` represents the name of the index to which the document will be added.
```

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
+- `<index>` represents the name of the index where the document will be added.
+- `<_id>` represents the unique identifier of the document.
+
+## Adding a sample index and documents to use the HTTP methods 
```
I'm not following the meaning of this heading. Please simplify/clarify.

leanneeliatra replied:
complete, this was confusing thank you.

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
+
+## Adding a sample index and documents to use the HTTP methods 
+
+To test the Document APIs, add a document by following the steps:
```
```suggestion
To test the Document APIs, add a document by following these steps:
```

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,52 @@ PUT <index>/_create/<_id>
+1. Open OpenSearch Dashboards.
+2. Navigate to the hamburger menu. 
+3. In the **Management** section, choose **Dev Tools**.
+4. Enter a command, then press the green triangle play button to send the request. Some example commands follow.
```
```suggestion
4. Enter a command, and then select the green triangle play button to send the request. The following are some example commands.
```

natebower commented on _security/access-control/document-level-security.md:
````diff
@@ -4,35 +4,35 @@ title: Document-level security
-   ```
-
-   - ![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)
+![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)
````
Please reference in the preceding text.

natebower commented on _security/access-control/document-level-security.md:
```diff
@@ -4,35 +4,35 @@ title: Document-level security
+- /security-plugin/access-control/document-level-security/
 ---
 
 # Document-level security (DLS)
```
```suggestion
# Document-level security
```

natebower commented on _security/access-control/document-level-security.md:
```diff
@@ -4,35 +4,35 @@ title: Document-level security
+
+## Simple roles
+
+Document-level security uses the OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document level security** section:
```
```suggestion
Document-level security uses OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document-level security** section:
```

natebower replied:
DSL should be defined on first appearance.

natebower reviewed (commented):
@leanneeliatra @Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

leanneeliatra commented:
> @leanneeliatra @Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

Thank you @natebower, I have addressed the editorial review comments and alos the comments from @Naarcha-AWS from our call.

Thank you both.

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+
+## Adding a sample index 
+
+Sample data can be added to the index with curl commands in the terminal, or through the API.
```
```suggestion
Sample data can be added to the index with curl commands in the terminal or through the API.
```

natebower commented on _api-reference/document-apis/index-document.md:
```diff
@@ -33,6 +33,53 @@ PUT <index>/_create/<_id>
+
+Sample data can be added to the index with curl commands in the terminal, or through the API.
+
+To test the Document APIs, add a document by following the steps:
```
```suggestion
To test the Document APIs, add a document by following these steps:
```

natebower commented on _security/access-control/document-level-security.md:
```diff
@@ -4,35 +4,35 @@ title: Document-level security
-
-   - ![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)
+# Document-level security
+Document-level security (DLS) lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index Permissions** section. See the Index Permissions section in OpenSearch Dashboards in the following image.
```
```suggestion
Document-level security lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index Permissions** section, shown in the following image.
```

natebower commented on _security/access-control/document-level-security.md:
```diff
@@ -4,35 +4,35 @@ title: Document-level security
+
+## Simple roles
+
+Document-level security uses OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document-level security** section:
```
```suggestion
Document-level security uses OpenSearch query domain-specific language (DSL) to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document-level security** section:
```

natebower commented on _security/access-control/document-level-security.md:
```diff
@@ -4,35 +4,35 @@ title: Document-level security
+
+This query specifies that for the role to have access to a document, its `genres` field must include `Comedy`.
+
+A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.
```
```suggestion
A typical request sent to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.
```

natebower reviewed (commented):
@leanneeliatra @Naarcha-AWS A few corrections, please. Thanks!

leanneeliatra commented:
> @leanneeliatra @Naarcha-AWS A few corrections, please. Thanks!

Brilliant @natebower thank you, those have been addressed.
